ext/app/client/HookStub.ts

```diff
@@ -81,3 +83,104 @@ function postEncode(options: {
   at.hash = url.hash;
   url.href = at.href;
 }
+
+/**
+ * Information about a download the user has requested.
+ */
+interface DownloadInfo {
+  data: any;     // The body of the response.
+  name: string;  // A name assigned in a header or elsewhere.
+  type: string;  // The mime type associated with the body.
+}
+
+/**
+ * Trigger a download of some prepared data in the browser.
+ * Must be called ultimately by a user clicking on a link
+ * or button, or the browser will deny a simulated click
+ * used in the implementation.
+ */
+async function downloadInBrowser(downloadInfo: DownloadInfo) {
+  const {data, name, type} = downloadInfo;
+  const blob = new Blob([data], {type});
+  const url = window.URL.createObjectURL(blob);
+  const link = document.createElement('a');
+  link.href = url;
+  link.download = name;
+  link.click();
+  window.URL.revokeObjectURL(url);
+}
+
+/**
+ * Given a URL of some document API REST endpoint, call
+ * the "back-end" code for it, and then try to determine
+ * the name and type (and content) of the response.
+ */
+async function fetchFromDocApi(href: string): Promise<DownloadInfo> {
+  // Figure out the path relative to the home URL
+  const config = getGristConfig();
+  if (!config?.homeUrl) {
+    // Not expected.
+    throw new Error('fetchFromDocApi needs homeUrl');
+  }
+  const pathname = '/' + href.split(config.homeUrl)[1];
+
+  // Use a simulated router to run the "back-end" code.
+  const app = gristOverrides.expressApp;
+  if (!app) {
+    // Not expected.
+    throw new Error('fetchFromDocApi needs expressApp');
+  }
+  const res = await app.run({
+    method: 'get',
+    path: pathname,
+  });
+
+  // Rifle through the information returned to find a name
+  // and type.
+  // A name may be provided in a Content-Disposition header
+  // of the form 'attachment; filename="name.csv"'
+  // or by a call to res.download
+  const prefix = 'attachment; filename="';
+  let name = String(res.headers?.['Content-Disposition'] || '');
+  if (name.startsWith(prefix)) {
+    name = name.slice(prefix.length, name.length - 1);
```

Comment:
```suggestion
    // Remove the prefix and the closing double quote
    name = name.slice(prefix.length, name.length - 1);
```

I was confused about the `- 1` for a bit

Reply:
that's the worst diff rendering i've ever seen